source/extensions/http/cache/file_system_http_cache/file_system_http_cache.cc

```diff
@@ -388,6 +405,78 @@ void FileSystemHttpCache::trackFileRemoved(uint64_t file_size) {
   stats_.size_bytes_.set(size_bytes_);
 }
 
+void FileSystemHttpCache::maybeEvict() {
+  uint64_t size_to_evict = 0;
+  uint64_t count_to_evict = 0;
+  if (config().has_max_cache_size_bytes()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_size = size_bytes_;
+    size_to_evict = seen_size > config().max_cache_size_bytes().value()
+                        ? seen_size - config().max_cache_size_bytes().value()
+                        : 0;
+  }
+  if (config().has_max_cache_entry_count()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_count = size_count_;
+    count_to_evict = seen_count > config().max_cache_entry_count().value()
+                         ? seen_count - config().max_cache_entry_count().value()
+                         : 0;
+  }
+  if (size_to_evict == 0 && count_to_evict == 0) {
+    return;
+  }
+
+  auto os_sys_calls = Api::OsSysCallsSingleton::get();
+  uint64_t size = 0;
+  uint64_t count = 0;
+  uint64_t proposed_size_evicted = 0;
+  struct ProposedEviction {
+    std::string name_;
+    uint64_t size_;
+    struct timespec last_touch_;
+    bool operator<(const ProposedEviction& other) const { return last_touch_ < other.last_touch_; }
+  };
+  std::multiset<ProposedEviction> proposed_evictions;
+
+  for (const Filesystem::DirectoryEntry& entry : Filesystem::Directory(std::string{cachePath()})) {
+    if (!isCacheFile(entry)) {
+      continue;
+    }
+    count++;
+    size += entry.size_bytes_.value_or(0);
+    struct stat s;
+    if (os_sys_calls.stat(absl::StrCat(cachePath(), entry.name_).c_str(), &s).return_value_ != -1) {
+      struct timespec last_touch = std::max(s.st_atim, s.st_ctim);
+      if (proposed_size_evicted < size_to_evict || proposed_evictions.size() < count_to_evict ||
+          last_touch < proposed_evictions.rbegin()->last_touch_) {
+        // We either haven't evicted enough yet, or this eviction candidate is 'older'
+        // than our current 'youngest' eviction candidate. So add this one to candidates.
+        proposed_evictions.insert(
+            ProposedEviction{entry.name_, entry.size_bytes_.value_or(0), last_touch});
+        proposed_size_evicted += entry.size_bytes_.value_or(0);
+        auto youngest = std::prev(proposed_evictions.end());
+        if (proposed_size_evicted - youngest->size_ >= size_to_evict &&
+            proposed_evictions.size() > count_to_evict) {
+          // We'd still be evicting enough if we don't evict the 'youngest' proposed eviction,
```

Comment:
I was expecting a loop here.

Let's say that I need to evict 1M to get to my target size, and the youngest 1000 files are 1k byte each, but then we find an older file that's 1M. We can un-propose evicting all the young 1k files.

Reply:
Agree. Looped.